sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySchemaIOProvider.java

```diff
@@ -0,0 +1,203 @@
+  }
+
+  /**
+   * Produces a SchemaIO given a String representing the data's location, the schema of the data
+   * that resides there, and some IO-specific configuration object.
+   *
+   * <p>For BigQuery IO, only the configuration object is used. Location and data schema have no
+   * effect. Specifying a table and dataset is done through appropriate fields in the configuration
+   * object, and the data schema is automatically generated from either the input PCollection or
+   * schema of the BigQuery table.
+   */
+  @Override
+  public BigQuerySchemaIO from(String location, Row configuration, @Nullable Schema dataSchema) {
+    return new BigQuerySchemaIO(location, configuration);
+  }
+
+  /**
+   * Indicates whether this transform requires a specified data schema.
+   *
+   * @return false
+   */
+  @Override
+  public boolean requiresDataSchema() {
+    return false;
+  }
+
+  /**
+   * Indicates whether the PCollections produced by this transform will contain a bounded or
+   * unbounded number of elements.
+   *
+   * @return Bounded
+   */
+  @Override
+  public PCollection.IsBounded isBounded() {
+    return PCollection.IsBounded.BOUNDED;
+  }
+
+  /** An abstraction to create schema aware IOs. */
+  static class BigQuerySchemaIO implements SchemaIO, Serializable {
+    protected final Row config;
+    protected final String location;
+
+    BigQuerySchemaIO(String location, Row config) {
+      this.config = config;
+      this.location = location;
+    }
+
+    /**
+     * Returns the schema of the data, if one was provided.
+     *
+     * @return null
+     */
+    @Override
+    @SuppressWarnings({
+      "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+    })
+    public Schema schema() {
+      return null;
+    }
+
+    /** Builds a schema aware reader via {@link BigQueryIO#readTableRowsWithSchema()}. */
+    @Override
+    public PTransform<PBegin, PCollection<Row>> buildReader() {
+      return new PTransform<PBegin, PCollection<Row>>() {
+        @Override
+        public PCollection<Row> expand(PBegin input) {
+          BigQueryIO.TypedRead<TableRow> read = BigQueryIO.readTableRowsWithSchema();
+          read = read.withMethod(BigQueryIO.TypedRead.Method.EXPORT);
+
+          String table = config.getString("table");
+          if (table != null) {
+            read = read.from(table);
+          }
+          String query = config.getString("query");
+          if (query != null) {
+            read = read.fromQuery(query).usingStandardSql();
+          }
+          String queryLocation = config.getString("queryLocation");
+          if (queryLocation != null) {
+            read = read.withQueryLocation(queryLocation);
+          }
+          return input.apply(read).apply(Convert.toRows());
+        }
+      };
+    }
+
+    /** Builds a schema aware writer via {@link BigQueryIO#write()}. */
+    @Override
+    public PTransform<PCollection<Row>, PDone> buildWriter() {
```

Comment:
Could we return a Row here to represent the WriteResult [1] instead ? (not sure if it works for the STORAGE_WRITE_API write method yet but if not we should be adding that in the future).

[1] https://github.com/apache/beam/blob/9418994e11dbede889b508c91261ec2092f2d8a7/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L1695

Reply:
This is probably approximately how we should be doing it. WriteResult is actually a collection of three PCollections as outputs: failedInserts, failedInsertsWithErr, and successfulInserts, with the second one being enabled only if a specific method is called (withExtendedErrorInfo).

So basically, it looks like the best approach here would be to add 2-3 outputs to the transform (depending on how many of those PCollections we want to support), and then convert all elements in the original WriteResults from TableRow to Beam Rows, with the built in conversion BigQuery IO currently uses. Doesn't sound too hard, but it does sound non-trivial enough to do in a separate PR.

Of course, the assumption there is that adding new outputs to the transform isn't a breaking change to users. Although I suppose in this case, "users" basically means us devs, which isn't that big a deal. I.e. adding WriteResults as outputs would mean modifying the Go and possibly Python wrappers, but otherwise wouldn't impact actual users of Beam...

Basically, I'm leaning towards a yes, but we should do this in future PRs. What do you think?

Reply:
I just remembered that I marked this experimental anyway. I added a warning about that in the documentation, but I think that's more reason to hold off for a followup PR.